### PR TITLE
Add support to rails 5.2

### LINKED ISSUE #26: Rails 5.2 alpha
I'm trying out edge Rails 5.2 alpha in a project that uses ransack. So this line fails:

https://github.com/activerecord-hackery/polyamorous/blob/2e4a587d4f3a8b074f2ee556781e94d835e15220/lib/polyamorous.rb#L41

```ruby
cannot load such file -- polyamorous/activerecord_5.2_ruby_2/join_association (LoadError)
```

What's your take on this? Should I just create those files? Copy paste the older ones? Should the conditional learn to simply load the latest version if none applicable can be found?

Thank you for your time!

### PR DESCRIPTION
**Pull request by gssbzn**
Initial work to bring support to rails 5.2

closes #26

### CODE REVIEW

**Comment by halo** [comment]
@gssbzn Do you feel this is mergable now? It sure seems to work. Are there any noteworthy differences to https://github.com/activerecord-hackery/polyamorous/pull/29?

/CC @seanfcarroll 
Thank you for your time

**Comment by seanfcarroll** [comment]
@bbonislawski  and @gssbzn given you have solved the same problem, do you think the two approaches should be merged? Alternatively vote for one of the approaches and then perhaps write some tests? I see there is a test in this PR, but I'd be happy seeing a few more.

That being said, there are a lot of people waiting on this in the Ransack and ActiveAdmin communities. If no-one is willing to spend time on tests it can be merged right away, but we might regrets that approach.

Thoughts?

**Comment by gssbzn** [comment]
@seanfcarroll @halo I did the fixes and changes based on the new 5.2 API for queries but wanted to test Ransack/ActiveAdmin to check nothing breaks. Sadly I've been busy at work this week but should have some free time this monday. If someone is able to check ransack referencing my branch over the weekend and provide some feedback I'll be thankful

**Comment by halo** [comment]
@gssbzn Thank you for your work. Are you not experiencing any of the problems listed in https://github.com/activerecord-hackery/polyamorous/issues/26#issuecomment-352079085? 

If I use `polyamorous` from your branch everything appears to work fine when I click around in the browser, but my specs don't pass at all because of several problems in `ransack` itself, which @bbonislawski has a fix for over here https://github.com/spark-solutions/ransack but I haven't tested that one yet.

@Petercopter can you provide a stacktrace for your `ActiveRecord::EagerLoadPolymorphicError` so we can find out whether `polyamorous` or `ransack` owns the problem?

**Comment by gssbzn** [comment]
@halo 

I fixed this:

```
ArgumentError:
        wrong number of arguments (given 3, expected 4)
```

As there was a change on the Rails API not addressed on #29, the rails team now pass an extra argument to track table aliases

I'll need to check @Petercopter error

**Comment by halo** [comment]
@seanfcarroll I suggest that you merge this PR for the following reasons:

- It's completely backwards compatible, so nothing in Rails 5.1 or earlier will break. The only code changes that touch 5.1 functionality have regression specs in this PR.
- It it blocking further/easier progress on the `ransack` gem itself because things break in `polyamorous` before they break anywhere else.
- It can't get worse, right now there is no support at all and people start using forks instead. This PR seems to work for the most part and may be fine-tuned later on.
- Much time has passed and I wouldn't want to lengthen that any more that because it might frustrate people trying to help battle-test Rails 5.2 or ActiveAdmin.

Thank you and Merry Christmas! ;)

**Comment by seanfcarroll** [comment]
Thanks for your efforts @halo ... fingers crossed!

**Comment by halo** [comment]
That was fast, haha! 😆 

Will you release it to Rubygems?

**Comment by seanfcarroll** [comment]
Yes, will do.

**Comment by seanfcarroll** [comment]
FYI - I'm just waiting on gem release permissions. Will be released ASAP.

File: lib/polyamorous/activerecord_5.2_ruby_2/join_association.rb [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+    # Reference: https://github.com/rails/rails/commit/9b15db5
+    # NOTE: Not sure we still need it?
+    #
+    def ==(other)
+      base_klass == other.base_klass
+    end
+
+    def build_constraint(klass, table, key, foreign_table, foreign_key)
+      if reflection.polymorphic?
```
Comment (kamipo) [inline lib/polyamorous/activerecord_5.2_ruby_2/join_association.rb:31]: Is this works as expected?
`JoinAssociation#build_constraint` was moved into `Reflection` at https://github.com/rails/rails/commit/442c15f1419ae60e4345866e5af59376a679e9d1.

**Pull request merged**